src/app/patient-table/patient-table.component.ts

```diff
@@ -16,17 +24,39 @@ import {
     templateUrl: './patient-table.component.html',
     styleUrls: ['./patient-table.component.css'],
 })
-export class PatientTableComponent {
+export class PatientTableComponent implements OnInit, OnChanges {
     @Output() updatePatient = new EventEmitter<Patient>();
+    @Input() newOrUpdatedPatient: Patient;
     dataSource!: Patient[];
     isLoadingResults = true;
 
     constructor(
         private patientService: PatientService,
         private snackService: SnackService,
         private route: Router,
-        public dialog: MatDialog,
+        private dialog: MatDialog,
     ) {}
+    ngOnChanges(changes: SimpleChanges): void {
+        if (changes['newOrUpdatedPatient'] && !!this.dataSource) {
+            const tablePatientIndex = this.dataSource.findIndex(
+                (ds: Patient) =>
+                    ds._id === changes['newOrUpdatedPatient'].currentValue._id,
+            );
+
+            if (tablePatientIndex !== -1) {
+                // update
+                this.dataSource[tablePatientIndex] =
+                    changes['newOrUpdatedPatient'].currentValue;
+                this.dataSource = [...this.dataSource];
+            } else {
+                // new
+                this.dataSource = [
+                    ...this.dataSource,
+                    changes['newOrUpdatedPatient'].currentValue,
+                ];
+            }
+        }
```

Comment:
do yo understand this part of the code?

Reply:
not yet but you can explain here

Reply:
This part is basically searching the index of the patient in the table data set

            const tablePatientIndex = this.dataSource.findIndex(
                (ds: Patient) =>
                    ds._id === changes['newOrUpdatedPatient'].currentValue._id,
            );

If index is !== -1 means that index was found the object exists so we are going to update it

           if (tablePatientIndex !== -1) {
                // update
                this.dataSource[tablePatientIndex] =
                    changes['newOrUpdatedPatient'].currentValue;
                this.dataSource = [...this.dataSource];
            }

other wise we do the insert as it seems to be a new patient added